sql/core/src/test/scala/org/apache/spark/sql/analysis/resolver/ExpressionIdAssignerSuite.scala

```diff
@@ -0,0 +1,818 @@
+            (originalProject.projectList, newProject.projectList, newProject)
+          } else {
+            (newProject.projectList, newProject)
+          }
+        case (originalOperator: LogicalPlan, newOperator: LogicalPlan) =>
+          if (originalOperator.resolved) {
+            (originalOperator.output, newOperator.output, newOperator)
+          } else {
+            (newOperator.output, newOperator)
+          }
+      }
+      .foreach {
+        case (
+            originalExpressions: Seq[_],
+            newExpressions: Seq[_],
+            newOperator: LogicalPlan
+            ) =>
+          originalExpressions.zip(newExpressions).zipWithIndex.foreach {
+            case (
+                (originalAttribute: AttributeReference, newAttribute: AttributeReference),
+                index
+                ) =>
+              if (leftmostOperators.containsKey(newOperator)) {
+                assert(
+                  originalAttribute.exprId == newAttribute.exprId,
+                  s"Attribute at $index was regenerated: $originalAttribute, $newAttribute"
+                )
+              } else {
+                assert(
+                  originalAttribute.exprId != newAttribute.exprId,
+                  s"Attribute at $index was not regenerated: $originalAttribute, $newAttribute"
+                )
+              }
+
+              attributesByName
+                .getOrElseUpdate(newAttribute.name, new ArrayBuffer[AttributeReference])
+                .append(newAttribute)
+            case ((originalAlias: Alias, newAlias: Alias), index) =>
+              if (leftmostOperators.containsKey(newOperator)) {
+                assert(
+                  originalAlias.exprId == newAlias.exprId,
+                  s"Alias at $index was regenerated: $originalAlias, $newAlias"
+                )
+              } else {
+                assert(
+                  originalAlias.exprId != newAlias.exprId,
+                  s"Alias at $index was not regenerated: $originalAlias, $newAlias"
+                )
+              }
+
+              aliasesByName.getOrElseUpdate(newAlias.name, new ArrayBuffer[Alias]).append(newAlias)
+          }
+        case (newExpressions: Seq[_], newOperator: LogicalPlan) =>
+          newExpressions.foreach {
+            case newAttribute: AttributeReference =>
+              attributesByName
+                .getOrElseUpdate(newAttribute.name, new ArrayBuffer[AttributeReference])
+                .append(newAttribute)
+            case newAlias: Alias =>
+              aliasesByName.getOrElseUpdate(newAlias.name, new ArrayBuffer[Alias]).append(newAlias)
+          }
+      }
+
+    attributesByName.values.foreach {
+      case attributes =>
+        val ids = attributes.map(attribute => attribute.exprId).distinct
+        assert(
+          ids.length == 1,
+          s"Different IDs for the same attribute in the plan: $attributes, $newPlan"
+        )
+    }
```

Comment:
You can convert to regular function call, I think:
```suggestion
    attributesByName.values.foreach { attributes =>
      val ids = attributes.map(attribute => attribute.exprId).distinct
      assert(
        ids.length == 1,
        s"Different IDs for the same attribute in the plan: $attributes, $newPlan"
      )
    }
```